File: Directory/154. Find Minimum in Rotatede Sorted Array II/Solution.py
# -*- coding: utf-8 -*-
# @Author: yufeng
# @Date:   2018-10-23 21:41:13
# @Last Modified by:   yufeng
# @Last Modified time: 2018-10-23 21:45:31
# @Email: fzhiy270@sina.com
# @Blog: http://fzhiy.com/

class Solution:
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        l, r = 0, len(nums)-1
        while l<r:
        	mid = (l+r)>>1
        	if nums[r]<nums[mid]:
        		l=mid+1
        	elif nums[r]>nums[mid]:
        		r=mid
        	else :
        		r-=1
        return nums[l]

class Solution:
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # return sorted(nums)[0]
        return sorted(list(set(nums)))[0]

File: Directory/154. Find Minimum in Rotatede Sorted Array II/Solution.cpp
/*
* @Author: yufeng
* @Date:   2018-10-23 21:33:42
* @Last Modified by:   yufeng
* @Last Modified time: 2018-10-23 21:40:56
* @Email: fzhiy270@sina.com
* @Blog: http://fzhiy.com/
*/

class Solution {
public:
    int findMin(vector<int>& nums) {
        int l = 0, r = nums.size()-1;
        while (l<r){
        	int mid=(l+r)>>1;
        	if(nums[r]<nums[mid]) l=mid+1;
        	else if(nums[r]>nums[mid]) r=mid;
        	else r--; /// 与153题不同的是 nums中可能有重复的数据，因为找最小的数，所以r--
        }
        return nums[l];
    }
};